webpack/scenes/ContentViews/__tests__/contentViewTable.test.js

```diff
@@ -0,0 +1,71 @@
+
+let firstCV;
+beforeEach(() => {
+  const { results } = contentViewIndex;
+  [firstCV] = results;
+});
+
+test('Can view content views on the screen', () => {
+  const { getByText, queryByTestId, queryByText } =
+    render(<ContentViewTable contentViews={contentViewIndex} />);
+
+  // query* functions will return the element or null if it cannot be found
+  // get* functions will return the element or throw an error if it cannot be found
+  // Loop through the CVs in the fixture and ensure they are present
+  contentViewIndex.results.map(({ name }) => {
+    expect(getByText(name)).toBeTruthy();
+    return name;
+  });
+
+
+  // Ensure loading text is not showing as a baseline sanity check
+  // Using .toBeNull() here as the loading spinner isn't even in the DOM, as opposed to
+  // an element that is in the DOM but hidden
+  // Using a test id that is set in the the application code so we can correctly identify
+  // the loading text. Pattern matching by "Loading" seems like it could give false positives, a
+  // test id will ensure we are correctly identifying the loading text.
+  expect(queryByTestId('cv-loading-text')).toBeNull();
+
+  // Ensure empty state text isn't showing as a baseline sanity check
+  expect(queryByText(/You currently don't have any Content Views/)).toBeFalsy();
+});
+
+test('Loading spinner is showing when no data is loaded yet', () => {
+  const { getByTestId } = render(<ContentViewTable contentViews={null} />);
+
+  // Now we check if the loading text is showing
+  expect(getByTestId('cv-loading-text')).toBeVisible();
+});
+
+test('Empty state message is shown when no Content Views are created yet', () => {
+  const noCvs = { results: [] };
+  const { getByText } = render(<ContentViewTable contentViews={noCvs} />);
+
+  expect(getByText(/You currently don't have any Content Views/)).toBeTruthy();
+});
+
+test('Can view Environment dropdown when cell is clicked', () => {
+  const { getByText, getByTestId } =
+    render(<ContentViewTable contentViews={contentViewIndex} />);
+
+  // Getting the row that corresponds with the first CV name
+  const firstRow = within(getByText(firstCV.name).closest('tr'));
+
+  // Make sure the environment expandable is not showing
+  // This uses a specific test id set in the application code, but as it is built out, it could use
+  // text shown on screen, which is prefered by the library. https://testing-library.com/docs/guide-which-query
+  // Using .toBeVisible() here as the element is in the DOM but hidden from the user's view
+  expect(getByTestId(`cv-environments-expansion-${firstCV.id}`)).not.toBeVisible();
+
+  // Click on the environments table cell in the row to expand the dropdown.
+  fireEvent.click(firstRow.getByLabelText(`environments-icon-${firstCV.id}`));
+
+  // Ensure the environment expandable is now visible on screen to the user.
+  expect(getByTestId(`cv-environments-expansion-${firstCV.id}`)).toBeVisible();
+});
```

Comment:
This file contains tests written with react-testing-library. I over-commented the file because of this since it is a new API and also as a reference example for later if we do use it more extensively.

I will write a community post about react-testing-library and open it up for more discussion. I've found it very intuitive and was able to write tests that resemble user interaction. I can see this being great for preventing bugs, writing maintainable tests, and doing TDD with bug fixes. I think it fits in with our UX-driven development for these pages as well.

Reply:
I also only wrote the test for the environments expandable and not the others as I want to wait until they are built out more before making assumptions.

There likely will be more test cases added overall, but its very early in development for this page and its hard to make certain assumptions just yet, also redux calls being added will likely complicate things. I tried to keep it to base functionality.

Reply:
discussion here! https://community.theforeman.org/t/using-react-testing-library-in-katello/18120